Copter: run rate loop at full filter rate in its own thread

andyp1per opened the pull request:
This is a redo of https://github.com/ArduPilot/ardupilot/pull/26189

I have squashed the commits, rebased and started fixing the underlying problems. There were some fundamental problems with how the original PR was handling attitude control changes so I thought it was better to just open a new PR.

Support is enabled by setting:
```
FSTRATE_ENABLE = 1
```
which gives a variable attitude rate depending on load, and:
```
FSTRATE_ENABLE = 2
```
which gives an attitude rate locked to the gyro rate, but dynamic while disarmed. For testing there is also:
```
FSTRATE_ENABLE = 3
```
which is a locked rate at all times.

The output rate can be adjusted using ```FSTRATE_DIV``` which is a gyro divisor (default 1).
Two different rates - and therefore tunes - can be compared by using the lua applet ```switch_rates.lua``` which will persistently switch the appropriate tune/rate parameters with ones saved in names beginning with ```X_```

IamPete1 commented on ArduCopter/Attitude.cpp:
```diff
@@ -4,18 +4,141 @@
+// @Field: dtMax: Max time delta since last log output
+// @Field: dtMin: Min time delta since last log output
+
+        if (now_us - last_rtdt_log_us >= 2500) {    // 400 Hz
+            AP::logger().WriteStreaming("RTDT", "TimeUS,dt,dtAvg,dtMax,dtMin", "Qffff",
```
```suggestion
        if (now_us - last_rtdt_log_us >= 2500) {    // 400 Hz
            last_rtdt_log_us = now_us;
            AP::logger().WriteStreaming("RTDT", "TimeUS,dt,dtAvg,dtMax,dtMin", "Qffff",
```

andyp1per replied:
Fixed

IamPete1 commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -275,6 +277,8 @@ float AC_PID::update_all(float target, float measurement, float dt, bool limit,
     _pid_info.D = D_out;
     _pid_info.FF = _target * _kff;
     _pid_info.DFF = _target_derivative * _kdff;
+    _pid_info.dt = dt;
+    _pid_info.max_dt = MAX(dt, _pid_info.dt);
```
This will presumably be the same for roll pitch and yaw? That is 4 redundant floats vs a time stamp and log header. New log message would use less log space especially if we ever wanted to expand logging of such things eg min dt.

andyp1per replied:
Yes agree, just WIP - will remove it when I have the bugs sorted

andyp1per replied:
Removed

andyp1per commented:
Flow today on top of 4.5 - working well.

andyp1per commented:
@rmackay9 this is close to being ready and I could use your input on what the configuration should look like. The basic premise is that we run the rate controller at the same rate as the gyros which is controlled by ```INS_GYRO_RATE``` (0=1Khz,1=2Khz,2=4Khz,3=8Khz). However, this quite expensive and will not be possible on some hardware setups. In addition we need to be able to log the rate outputs at a faster rate than the main loop rate, but its unlikely that we could successfully log at the gyro rate. Finally we need to update the filters (notch filter in particular) at a fast rate than the main loop, but it does not make sense to do this any faster than at half the gyro rate. 

Currently we have the following:
```FLIGHT_OPTIONS=8``` - enables the rate thread and dynamically schedules the relationship to the gyro rate based on load
```FLIGHT_OPTIONS=16``` - fixes the rate thread at the same rate as the gyro rate
```ATT_FILTER_RATE``` - update rate of the filters in Hz or fixed at half the gyro rate if set to 0
```ATT_LOG_RATE``` - update rate of the rate logs in Hz or set at the main loop rate if set to 0
The last two are not exact numbers - they are just the closest we can get to the required frequency by counting gyro samples so need to be an integer divisor or the gyro rate.

I'm not currently convinced about the naming or the configuration. I think this is an important enough feature that it needs its own enable flag. I also think it would be worth having a rate parameter that allows you to fix the rate. I also think that ATT is a little bit misleading. Maybe FAST_RATE_ or RATET_ or something. So an alternative might be:

```FRATE_ENABLE``` - set to 1 to enable the rate thread
```FRATE_RATE``` - set to 0 to dynamically throttle, fixed Hz otherwise
```FRATE_LOG_RATE``` - as above
```FRATE_FILT_RATE``` - as above

or something. It would also be possible to not use Hz at all but simply pick the divisor. So 2 would be gyro rate / 2 etc.
Your thoughts appreciated on this important topic!

IamPete1 commented on ArduCopter/Attitude.cpp:
```diff
@@ -1,21 +1,285 @@
+                // setup the notch filter sample rate
+                const float loop_rate_hz = AP::scheduler().get_filtered_loop_rate_hz();
+                attitude_control->set_notch_sample_rate(loop_rate_hz);
+                motors->set_dt(1.0/loop_rate_hz);
```
motors dt is set lots of places, but I don't think the motors update rate is actually changed?

andyp1per replied:
Fixed

IamPete1 reviewed (commented):
It would be good to split this out into several smaller PRs. Things like the DAL rename would stand alone?

andyp1per commented:
Absolutely, but I want to make sure there is general agreement on the structure before doing thatAndySent from my iPhoneOn 16 Jun 2024, at 11:36, Peter Hall ***@***.***> wrote:﻿
@IamPete1 commented on this pull request.

It would be good to split this out into several smaller PRs. Things like the DAL rename would stand alone?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

peterbarker commented on Tools/autotest/arducopter.py:
```diff
@@ -6183,6 +6183,77 @@ def DynamicRpmNotches(self):
+            "SIM_ESC_TELEM": 1,
+            "FLIGHT_OPTIONS": 8
+        })
+        self.reboot_sitl()
```
We need a `context_push()` at the top of this test and `context_pop()` at the bottom followed by the existing `reboot_sitl()`

... otherwise subsequent tests will end up running with the wrong parameters in effect (not value, effect).  Assuming this reboot_sitl() is here for a reason....

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_HAL/board/chibios.h:
```diff
@@ -86,6 +88,10 @@
 #define HAL_INS_ICM20608_AM_NAME "icm20608-am"
 #define HAL_INS_ICM20608_EXT_NAME "icm20608_ext"
 
+#ifndef HAL_INS_RATE_LOOP
+#define HAL_INS_RATE_LOOP 1
+#endif
+
```
Maybe mve this default-to-1 code up into `AP_HAL/AP_HAL_Boards.h`?

andyp1per replied:
Fixed

andyp1per commented:
Did some fairly comprehensive flying including checking that I could switch between rates mid-flight - all worked fine

tridge commented on ArduCopter/Attitude.cpp:
```diff
@@ -1,21 +1,345 @@
-// update rate controllers and output to roll, pitch and yaw actuators
-//  called at 400hz by default
-void Copter::run_rate_controller()
+#if AP_INERTIALSENSOR_RATE_LOOP_WINDOW_ENABLED
```
maybe break into fast_rate.cpp ?

andyp1per replied:
done

tridge reviewed (changes requested):
my preference for 4.6.x would be this as a custom build server feature, but should ask @rmackay9 about this.
It should have less or preferably zero flash impact on plane.
I'd like more info on why condition variables are needed and why binary semaphores don't work as well, preferably a version of the binary semaphore test that shows the issue

andyp1per commented:
This is the rough cost for the refactoring required:

```
andy@Eagle:~/github/ardupilot$ ./Tools/scripts/pretty_diff_sym.py ../ardupilot-dev/build/MatekH743-bdshot/bin/arduplane build/MatekH743-bdshot/bin/arduplane
Symbol size differences between ../ardupilot-dev/build/MatekH743-bdshot/bin/arduplane and build/MatekH743-bdshot/bin/arduplane:
  18 AP_InertialSensor_Backend::_publish_gyro(unsigned 126 -> 144
   4 AC_AttitudeControl::input_thrust_vector_heading(Vector3<float> 376 -> 380
 180 AP_InertialSensor_Backend::update_gyro_filters(unsigned 0 -> 180
-190 AC_AttitudeControl::input_rate_bf_roll_pitch_yaw_2(float, 212 -> 22
   2 AP_HAL::RCOutput::force_trigger_groups(bool) 0 -> 2
   4 ChibiOS::DeviceBus::bus_thread(void*) 188 -> 192
   6 ChibiOS::SPIDevice::set_periodic_minimum(unsigned 0 -> 6
  -8 ChibiOS::RCOutput::push_local() 408 -> 400
  20 chMtxLockS 120 -> 140
   4 AP_AHRS_get_accel(lua_State*) 84 -> 88
   4 AC_AttitudeControl::input_euler_angle_roll_pitch_yaw(float, 540 -> 544
   2 AP_InertialSensor_Backend::set_primary_gyro(unsigned 0 -> 2
  -8 AC_AttitudeControl::attitude_controller_run_quat() 508 -> 500
   4 AC_AttitudeControl::relax_attitude_controllers() 176 -> 180
   4 SLT_Transition::update() 1148 -> 1152
   4 AC_AttitudeControl::input_rate_bf_roll_pitch_yaw(float, 296 -> 300
   4 AC_AttitudeControl_TS::input_euler_rate_yaw_euler_angle_pitch_bf_roll(bool, 556 -> 560
  20 AP_AHRS_View::Write_Rate(AP_Motors 484 -> 504
   8 SRV_Channels::calc_pwm() 212 -> 220
   4 ChibiOS::DeviceBus::DeviceBus(unsigned 44 -> 48
   8 rcoutDriver 672 -> 680
  44 SRV_Channels::output_pwm_chan(unsigned 0 -> 44
  16 AC_AttitudeControl::input_rate_bf_roll_pitch_yaw_3(float, 472 -> 488
  24 AP_InertialSensor::update() 464 -> 488
   4 AC_AttitudeControl::input_angle_step_bf_roll_pitch_yaw(float, 136 -> 140
  -4 AP_InertialSensor::write_notch_log_messages() 368 -> 364
   4 AC_AttitudeControl::input_euler_angle_roll_pitch_euler_rate_yaw(float, 420 -> 424
   4 ChibiOS::RCOutput::set_dshot_rate(unsigned 136 -> 140
   6 ChibiOS::I2CDevice::set_periodic_minimum(unsigned 0 -> 6
  24 AP_Motors::add_motor_num(signed 26 -> 50
  12 AC_AttitudeControl_Multi::AC_AttitudeControl_Multi(AP_AHRS_View&, 480 -> 492
  62 AP_MotorsMulticopter::update_servo_map() 0 -> 62
   8 AP_Vehicle::update_dynamic_notch_at_specified_rate() 92 -> 100
 300 AC_AttitudeControl_Multi::rate_controller_run_dt(float, 0 -> 300
 104 AC_AttitudeControl::input_rate_step_bf_roll_pitch_yaw(float, 0 -> 104
   4 AP_AHRS::_get_primary_accel_index() 0 -> 4
   4 AC_AttitudeControl::input_euler_rate_roll_pitch_yaw(float, 384 -> 388
   4 GCS_MAVLINK_Plane::send_attitude_target() 104 -> 108
   2 AP_Motors::rc_write(unsigned 72 -> 74
 -32 AP_InertialSensor_Backend::update_accel(unsigned 134 -> 102
   6 AP_AHRS::write_video_stabilisation() 186 -> 192
  28 CSWTCH.378 0 -> 28
   2 AP_InertialSensor_Backend::set_primary_accel(unsigned 0 -> 2
   2 AC_AttitudeControl::rate_controller_run_dt(float, 0 -> 2
   4 AP_MotorsMulticopter::output() 80 -> 84
  60 AP_InertialSensor_Backend::update_accel_filters(unsigned 0 -> 60
 228 AC_AttitudeControl::input_rate_bf_roll_pitch_yaw_4(float, 0 -> 228
 -64 AC_AutoTune_Multi::twitch_test_run(AC_AutoTune::AxisType, 832 -> 768
-290 AC_AttitudeControl_Multi::rate_controller_run() 332 -> 42
   8 ChibiOS::I2CDeviceManager::businfo 200 -> 208
   4 AP_InertialSensor_Invensense::start() 924 -> 928
   8 ChibiOS::RCOutput::trigger_groups() 204 -> 212
  44 AC_AttitudeControl_Multi::rate_controller_target_reset() 0 -> 44
 -28 CSWTCH.368 28 -> 0
   8 Plane::set_takeoff_expected() 140 -> 148
  48 AP_InertialSensor_Invensense::_read_fifo() 524 -> 572
   6 ChibiOS::RCOutput::force_trigger_groups(bool) 0 -> 6
   2 AP_InertialSensor_Invensensev3::set_primary_gyro(unsigned 0 -> 2
  -6 AP_InertialSensor_Backend::apply_gyro_filters(unsigned 276 -> 270
-172 AP_InertialSensor_Backend::update_gyro(unsigned 274 -> 102
   2 AC_AttitudeControl::rate_controller_target_reset() 0 -> 2
   2 AP_InertialSensor_Invensense::set_primary_gyro(unsigned 0 -> 2
Total: 586
```

andyp1per commented:
The reason why BinarySemaphore does not work in this case is two-fold:

1. It does not use mutexes and so there is no priority inheritance between related threads, this was one reason why the original code had to increase the thread prioty of the attitude thread. The new code has no need of this - the priorities manage themselves. This makes for a safer change.
2. You if you wait after signal has been called you will timeout. This is because binary semaphores contain no history - there is no matching of wait to signal. This is easy to verify by making the first thread in the test signal and delay before creating the second thread. We are managing a queue of gyro samples of which we need to process all. If a gyro sample is queued before the attitude thread is ready to receive then signal is lost and the attitude thread will not get woken up until the *next* gyro sample arrives. Obviously not good. To try and fix this you could then have an additional mutex to control access to the number of waiting gyros, but at that point you have essentially implemented condition variables which ChibiOS has a much more low-level and efficient implementation of - plus you still don't get the benefits of point (1).

tridge commented:
I hope I'm not just being thick here, but I think I really do need to understand what is going on here. Potentially dumb questions below.

> 1. It does not use mutexes and so there is no priority inheritance between related threads, this was one reason why the original code had to increase the thread prioty of the attitude thread. The new code has no need of this - the priorities manage themselves. This makes for a safer change.

first off, by "attitude thread" I assume you mean the rate thread? The attitude control is still in the main thread as far as I can see, only rate update in the rate thread. I know we discussed moving the attitude update into that thread but unless I've missed it that isn't being done in this PR, and it would be a tricky thing to do as you would need an attitude update (so would need to either run EKF propagation at high rate or do your own propagation using gyro/accel data)
secondly, I can't see how priority inheritance plays any role at all. The only two threads involved with this CondMutex are the SPI threads doing the transfers and filtering and the rate thread. Both are at priority 181, so neither change priority at all. Is there some interaction I'm missing that causes priority inheritance to kick in? If so, between what threads and under what circumstances?

> 2\. This is easy to verify by making the first thread in the test signal and delay before creating the second thread.

the signal can come from any thread, the state is in the binary semaphore object itself, so I don't see how the first thread delaying makes a difference. I'd like to have this as an actual test to understand it

tridge commented:
> 2\. This is easy to verify by making the first thread in the test signal and delay before creating the second thread.

I've updated the BinarySem test to do what you describe and it works and behaves exactly as I expected

https://github.com/tridge/ardupilot/commits/pr-copter-rate-thread/

I used a blocking wait, which matches what you are doing in this PR, no timeouts (you can't timeout with a blocking wait, it blocks forever), and gets exactly the performance I expect:

```
tick 21068 99.681 ops/s 0.000 timeouts/s
tick 22071 99.681 ops/s 0.000 timeouts/s
tick 23075 99.681 ops/s 0.000 timeouts/s
tick 24078 99.681 ops/s 0.000 timeouts/s
tick 25081 99.681 ops/s 0.000 timeouts/s
tick 26084 99.681 ops/s 0.000 timeouts/s
tick 27087 99.681 ops/s 0.000 timeouts/s
tick 28091 99.681 ops/s 0.000 timeouts/s
tick 29094 99.681 ops/s 0.000 timeouts/s
tick 30097 99.681 ops/s 0.000 timeouts/s
tick 31100 99.681 ops/s 0.000 timeouts/s
tick 32103 99.681 ops/s 0.000 timeouts/s
tick 33107 99.681 ops/s 0.000 timeouts/s
```

tridge commented:
> 1. It does not use mutexes and so there is no priority inheritance between related threads, this was one reason why the original code had to increase the thread prioty of the attitude thread. The new code has no need of this - the priorities manage themselves. This makes for a safer change.

I just noticed that the rate thread is being created with AP_HAL::Scheduler::PRIORITY_RCOUT+1, so it is priority 182, and SPI threads are priority 181, so there are different priorities, but the rate thread is the one waiting, and priority inheritance only ever boosts priority of the waiting thread. As the thread that is waiting (the rate thread) is at a higher priority than the thread sending the signal there can be no priority change, so I'm still puzzled as to how priority inheritance comes into the picture at all.

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -257,6 +272,28 @@ void AP_InertialSensor_Backend::apply_gyro_filters(const uint8_t instance, const
+          tell the rate thread we have a new sample
+        */
+        if (++_imu.rate_decimation_count >= _imu.rate_decimation) {
+            _imu._cmutex->lock_and_signal();
```
why are you signalling before you push?? I think this may be why you are getting the odd binary sem behaviour. You have to signal after you have made the sample available or the rate thread will wake up pointlessly and you will always be one loop behind

andyp1per replied:
This makes no difference - the waiter will not be woken up until it gets the mutex which won't happen until the mutex is unlocked by the signaler.

If the waiter is in chCondWait the first thing it needs to do is re-aquire the mutex which it cannot do until the signaler has released it.

https://chibios.org/dokuwiki/lib/exe/fetch.php?media=chibios:documentation:books:rt:kernel:monitor_states.png

As the docs state this is very similar to java synchronized which I am very familiar with:

https://chibios.org/dokuwiki/doku.php?id=chibios:documentation:books:rt:kernel_mutexes

tridge commented:
@andyp1per we had a "climb away" event with this PR on Sunday with Michelle's quad, with FSTRATE_ENABLE=2. Looks like Z controller was not updating in ALT_HOLD mode.

andyp1per commented:
> I've updated the BinarySem test to do what you describe and it works and behaves exactly as I expected

Ok I have added an example that I think illustrates the problem. With binary sem:
```
Setup threads
tick 1000 1659.794 ops/s 0.000 timeouts/s missing 13
tick 2000 1700.771 ops/s 0.000 timeouts/s missing 22
tick 3000 1586.266 ops/s 0.000 timeouts/s missing 13
tick 4001 1695.454 ops/s 0.000 timeouts/s missing 17
tick 5001 1641.411 ops/s 0.000 timeouts/s missing 23
tick 6001 1783.319 ops/s 0.000 timeouts/s missing 11
tick 7001 1772.874 ops/s 0.000 timeouts/s missing 17
```
with CondMutex:
```
Setup threads
tick 1000 1692.284 ops/s 0.000 timeouts/s missing 0
tick 2000 1685.213 ops/s 0.000 timeouts/s missing 0
tick 3001 1574.518 ops/s 0.000 timeouts/s missing 0
tick 4001 1615.891 ops/s 0.000 timeouts/s missing 0
tick 5001 1660.900 ops/s 0.000 timeouts/s missing 0
tick 6002 1733.622 ops/s 0.000 timeouts/s missing 0
```
it's obviously a bit contrived, but the point is that BinarySemphore will not work on its own and actually from a concurrency POV I am not sure it can be made to work at all because it is not a lock in the true sense of the word - you end up with all these narrow races if you start trying to use a mutex to manage a counter whereas CondMutex does not suffer from this problem because the check is completely atomic with the wait/signal. I guess one of the issues here is that I moved on from your original code that was only dealing with a single sample onto this little buffer of samples that makes the concurrency more tricky.

andyp1per commented:
You are correct about the name - I have been using "attitude" and "rate" interchangeably when in fact it is rate. In my defence it was all inside a file called "Attitude.cpp". I have changed the name of the thread to reflect this.

tridge commented:
> Ok I have added an example that I think illustrates the problem

in that example I don't see why you expect missing to be zero. With the random sleeps it could easily be that thread1 calls signal several times while thread2 is sleeping. You only want thread2 to get one event in that case, so "sent - ops" with be non-zero with quite a high chance. 
Also, get_random() returns 0 one in 256 times, so that means one in 256 times thread1 sends multiple signals with no chance for thread2 to run at all as thread1 doesn't sleep.

tridge commented:
> I guess one of the issues here is that I moved on from your original code that was only dealing with a single sample onto this little buffer of samples that makes the concurrency more tricky.

but you do process that whole buffer at once when you wake up, yes? If so then if that buffer ever gets more than one entry in it then you don't want to wake up every time.

I've pushed a fix for your test and it works for BinarySem. 
https://github.com/tridge/ardupilot/commits/pr-copter-rate-thread/

andyp1per commented:
> > I guess one of the issues here is that I moved on from your original code that was only dealing with a single sample onto this little buffer of samples that makes the concurrency more tricky.
> 
> but you do process that whole buffer at once when you wake up, yes? If so then if that buffer ever gets more than one entry in it then you don't want to wake up every time.
> 
> I've pushed a fix for your test and it works for BinarySem. https://github.com/tridge/ardupilot/commits/pr-copter-rate-thread/

This ignores the concurrency issue I was alluding to - the update to bsize is not atomic with the wait/notification. It works because in the test the race is so narrow - but its there and is more obvious when you need to do something more complicated than simply update a counter.

andyp1per commented:
Now only compiled in on Copter, still some cost on other platforms due to the minor refactorings:

```
andy@Eagle:~/github/ardupilot$ ./Tools/scripts/pretty_diff_sym.py ../ardupilot-dev/build/MatekH743-bdshot/bin/arduplane ./build/MatekH743-bdshot/bin/arduplane | sort --key=2
Total: 628
  -8 AC_AttitudeControl::attitude_controller_run_quat() 508 -> 500
   4 AC_AttitudeControl::input_angle_step_bf_roll_pitch_yaw(float, 136 -> 140
   4 AC_AttitudeControl::input_euler_angle_roll_pitch_euler_rate_yaw(float, 420 -> 424
   4 AC_AttitudeControl::input_euler_angle_roll_pitch_yaw(float, 540 -> 544
   4 AC_AttitudeControl::input_euler_rate_roll_pitch_yaw(float, 384 -> 388
-190 AC_AttitudeControl::input_rate_bf_roll_pitch_yaw_2(float, 212 -> 22
  16 AC_AttitudeControl::input_rate_bf_roll_pitch_yaw_3(float, 472 -> 488
 228 AC_AttitudeControl::input_rate_bf_roll_pitch_yaw_4(float, 0 -> 228
   4 AC_AttitudeControl::input_rate_bf_roll_pitch_yaw(float, 296 -> 300
 104 AC_AttitudeControl::input_rate_step_bf_roll_pitch_yaw(float, 0 -> 104
   4 AC_AttitudeControl::input_thrust_vector_heading(Vector3<float> 376 -> 380
  12 AC_AttitudeControl_Multi::AC_AttitudeControl_Multi(AP_AHRS_View&, 480 -> 492
-290 AC_AttitudeControl_Multi::rate_controller_run() 332 -> 42
 300 AC_AttitudeControl_Multi::rate_controller_run_dt(float, 0 -> 300
  44 AC_AttitudeControl_Multi::rate_controller_target_reset() 0 -> 44
   2 AC_AttitudeControl::rate_controller_run_dt(float, 0 -> 2
   2 AC_AttitudeControl::rate_controller_target_reset() 0 -> 2
   4 AC_AttitudeControl::relax_attitude_controllers() 176 -> 180
   4 AC_AttitudeControl_TS::input_euler_rate_yaw_euler_angle_pitch_bf_roll(bool, 556 -> 560
 -64 AC_AutoTune_Multi::twitch_test_run(AC_AutoTune::AxisType, 832 -> 768
   4 AP_AHRS_get_accel(lua_State*) 84 -> 88
   4 AP_AHRS::_get_primary_accel_index() 0 -> 4
  20 AP_AHRS_View::Write_Rate(AP_Motors 484 -> 504
   6 AP_AHRS::write_video_stabilisation() 186 -> 192
   2 AP_HAL::RCOutput::force_trigger_groups(bool) 0 -> 2
   2 AP_InertialSensor_Backend::apply_gyro_filters(unsigned 276 -> 278
  40 AP_InertialSensor_Backend::_publish_gyro(unsigned 126 -> 166
   2 AP_InertialSensor_Backend::set_primary_accel(unsigned 0 -> 2
   2 AP_InertialSensor_Backend::set_primary_gyro(unsigned 0 -> 2
  60 AP_InertialSensor_Backend::update_accel_filters(unsigned 0 -> 60
 -32 AP_InertialSensor_Backend::update_accel(unsigned 134 -> 102
 180 AP_InertialSensor_Backend::update_gyro_filters(unsigned 0 -> 180
-172 AP_InertialSensor_Backend::update_gyro(unsigned 274 -> 102
  48 AP_InertialSensor_Invensense::_read_fifo() 524 -> 572
   2 AP_InertialSensor_Invensense::set_primary_gyro(unsigned 0 -> 2
   4 AP_InertialSensor_Invensense::start() 924 -> 928
   2 AP_InertialSensor_Invensensev3::set_primary_gyro(unsigned 0 -> 2
   2 AP_InertialSensor::push_next_gyro_sample(unsigned 0 -> 2
   4 AP_InertialSensor::push_rate_loop_gyro(unsigned 0 -> 4
  24 AP_InertialSensor::update() 464 -> 488
  -4 AP_InertialSensor::write_notch_log_messages() 368 -> 364
  24 AP_Motors::add_motor_num(signed 26 -> 50
   4 AP_MotorsMulticopter::output() 82 -> 86
  62 AP_MotorsMulticopter::update_servo_map() 0 -> 62
   2 AP_Motors::rc_write(unsigned 72 -> 74
   8 AP_Vehicle::AP_Vehicle() 468 -> 476
   4 AP_Vehicle::one_Hz_update() 44 -> 48
   8 AP_Vehicle::setup() 664 -> 672
   8 AP_Vehicle::update_dynamic_notch_at_specified_rate() 92 -> 100
   4 ChibiOS::DeviceBus::bus_thread(void*) 188 -> 192
   4 ChibiOS::DeviceBus::DeviceBus(unsigned 44 -> 48
   8 ChibiOS::I2CDeviceManager::businfo 200 -> 208
   6 ChibiOS::I2CDevice::set_periodic_minimum(unsigned 0 -> 6
   6 ChibiOS::RCOutput::force_trigger_groups(bool) 0 -> 6
  -8 ChibiOS::RCOutput::push_local() 408 -> 400
   4 ChibiOS::RCOutput::set_dshot_rate(unsigned 136 -> 140
   8 ChibiOS::RCOutput::trigger_groups() 204 -> 212
   6 ChibiOS::SPIDevice::set_periodic_minimum(unsigned 0 -> 6
  20 chMtxLockS 120 -> 140
 -28 CSWTCH.368 28 -> 0
  28 CSWTCH.378 0 -> 28
   4 GCS_MAVLINK_Plane::send_attitude_target() 104 -> 108
  -4 GCS_MAVLINK_Plane::send_nav_controller_output() 648 -> 644
  -8 ModeAcro::stabilize() 484 -> 476
  -4 ModeLoiterAltQLand::_enter() 144 -> 140
  -4 ModeQLoiter::run() 928 -> 924
  -4 ModeQRTL::run() 748 -> 744
 -16 ModeThermal::exit_heading_aligned() 164 -> 148
   8 plane 37520 -> 37528
  -4 Plane::ahrs_update() 252 -> 248
  -4 Plane::apply_throttle_limits(float) 336 -> 332
   2 Plane::avoidance_adsb_update() 26 -> 28
   8 Plane::load_parameters() 928 -> 936
   4 Plane::Log_Write_AETR() 156 -> 160
   4 Plane::Log_Write_Attitude() 416 -> 420
   4 Plane::Log_Write_Control_Tuning() 232 -> 236
   4 Plane::Log_Write_Guided() 100 -> 104
   4 Plane::Log_Write_Nav_Tuning() 216 -> 220
   4 Plane::Log_Write_OFG_Guided() 148 -> 152
   2 Plane::Log_Write_RC() 48 -> 50
   4 Plane::Log_Write_Status() 148 -> 152
   4 Plane::Log_Write_Vehicle_Startup_Messages() 100 -> 104
   2 Plane::mode_from_mode_num(Mode::Number) 220 -> 222
  -4 Plane::notify_mode(Mode 48 -> 44
   4 Plane::one_second_loop() 520 -> 524
  -8 Plane::Plane() 1940 -> 1932
   8 Plane::set_takeoff_expected() 140 -> 148
   2 Plane::should_log(unsigned 8 -> 10
   4 Plane::stabilize_stick_mixing_fbw() 508 -> 512
   4 Plane::start_command(AP_Mission::Mission_Command 976 -> 980
  -4 Plane::takeoff_calc_throttle(bool) 148 -> 144
  -4 Plane::trigger_land_abort(float) 224 -> 220
  -4 Plane::update_alt() 444 -> 440
  -2 Plane::update_speed_height() 74 -> 72
   4 QuadPlane::Log_Write_QControl_Tuning() 384 -> 388
  -8 QuadPlane::verify_vtol_land() 688 -> 680
   8 rcoutDriver 672 -> 680
   8 SRV_Channels::calc_pwm() 212 -> 220
  44 SRV_Channels::output_pwm_chan(unsigned 0 -> 44
```

andyp1per commented on Tools/scripts/extract_features.py:
```diff
@@ -196,7 +196,8 @@ def __init__(self, filename, nm="arm-none-eabi-nm", strings="strings"):
 
             ('HAL_WITH_DSP', r'AP_HAL::DSP::find_peaks\b',),
-            ('AP_INERTIALSENSOR_HARMONICNOTCH_ENABLED', r'AP_InertialSensor::HarmonicNotch::update_params\b',),
+            ('AP_INERTIALSENSOR_HARMONICNOTCH_ENABLED', r'AP_InertialSensor::get_next_gyro_sample\b',),
```
fixed

andyp1per commented:
Ok done another update to my example. I *think* it shows what I am talking about - and is at least consistent with what I was observing - noticeably that with binary semaphore you get some inconsistency and delay in processing samples:

BSem:
```
tick 6000 3932.320 ops/s missing 0 max_delay 278us
tick 7000 3931.796 ops/s missing -1 max_delay 280us
tick 8000 3932.532 ops/s missing -1 max_delay 277us
tick 9001 3929.536 ops/s missing -1 max_delay 277us
tick 10001 3937.906 ops/s missing 0 max_delay 277us
tick 11001 3931.693 ops/s missing -1 max_delay 277us
tick 12001 3930.709 ops/s missing -1 max_delay 280us
tick 13001 3928.525 ops/s missing -1 max_delay 329us
tick 14001 3929.309 ops/s missing -1 max_delay 277us
tick 15001 3931.921 ops/s missing -1 max_delay 278us
tick 16001 3933.913 ops/s missing -1 max_delay 276us
tick 17001 3930.422 ops/s missing -1 max_delay 278us
tick 18001 3932.618 ops/s missing 0 max_delay 277us
tick 19001 3934.658 ops/s missing 0 max_delay 276us
```
CondMutex:
```
tick 1000 3907.441 ops/s missing 0 max_delay 121us
tick 2000 3918.130 ops/s missing 0 max_delay 123us
tick 3000 3915.374 ops/s missing 0 max_delay 123us
tick 4000 3914.440 ops/s missing 0 max_delay 120us
tick 5000 3913.452 ops/s missing 0 max_delay 125us
tick 6000 3914.499 ops/s missing 0 max_delay 123us
tick 7001 3913.385 ops/s missing 0 max_delay 118us
tick 8001 3913.288 ops/s missing 0 max_delay 120us
tick 9001 3913.362 ops/s missing 0 max_delay 121us
```
This is running on an actual H7 with ChibiOS

andyp1per commented:
I think a good chunk of the size changes are alignment changes - not sure I can do anything about that

andyp1per commented:
Re-flown with latest changes - all good

tridge commented:
why the cost on plane on disables board?
![image](https://github.com/user-attachments/assets/991af9a6-6a1e-45e8-be43-7ae4d2ac0d7c)

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor_Invensense.cpp:
```diff
@@ -487,6 +488,16 @@ bool AP_InertialSensor_Invensense::update() /* front end */
+    if (!_imu.use_rate_loop_gyro_samples()) {
+        return;
+    }
+    _dev->set_periodic_minimum(instance == gyro_instance ? 0 : 100);
```
this seems really dangerous, the SPI thread could completely dominate the CPU

andyp1per commented:
@tridge I've split out the safe + necessary PR's for separate review to make the unsafe + scary stuff easier to review and control at compile time.

andyp1per commented:
This is now a 100byte cost on plane, which is mainly due to the small restructuring of INS:
```
Binary Name      Text [B]         Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ---------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   104 (+0.0077%)   0 (0.0000%)  0 (0.0000%)    104 (+0.0077%)                                   621084
arducopter       3400 (+0.1877%)  0 (0.0000%)  0 (0.0000%)    3400 (+0.1872%)                                  146872
ardusub          104 (+0.0065%)   0 (0.0000%)  0 (0.0000%)    104 (+0.0065%)                                   359460
blimp            104 (+0.0076%)   0 (0.0000%)  0 (0.0000%)    104 (+0.0076%)                                   599032
arduplane        116 (+0.0064%)   0 (0.0000%)  4 (+0.0015%)   116 (+0.0064%)                                   152832
ardurover        108 (+0.0065%)   0 (0.0000%)  -4 (-0.0015%)  108 (+0.0065%)                                   302264
arducopter-heli  3404 (+0.1880%)  0 (0.0000%)  -4 (-0.0015%)  3404 (+0.1876%)                                  147768
```

andyp1per commented on libraries/AP_InertialSensor/FastRateBuffer.cpp:
```diff
@@ -0,0 +1,124 @@
+{
+    return fast_rate_buffer->get_num_gyro_samples();
+}
+
```
add comments

andyp1per replied:
Fixed

andyp1per commented on libraries/AP_InertialSensor/FastRateBuffer.cpp:
```diff
@@ -0,0 +1,124 @@
+
+uint32_t AP_InertialSensor::get_num_gyro_samples()
+{
+    return fast_rate_buffer->get_num_gyro_samples();
```
check for nullptr?

andyp1per replied:
Fixed

rmackay9 commented on libraries/AP_InertialSensor/FastRateBuffer.cpp:
```diff
@@ -0,0 +1,124 @@
+        _notifier.wait_blocking();
+    }
+
+    WITH_SEMAPHORE(_mutex);
```
I always thought this had to be within curly braces or else it was equivalent to having it at the top of the method

andyp1per commented on libraries/AP_HAL_ChibiOS/RCOutput.h:
```diff
@@ -579,6 +584,8 @@ class ChibiOS::RCOutput : public AP_HAL::RCOutput
     uint8_t _dshot_cycle;
     // virtual timer for post-push() pulses
     virtual_timer_t _dshot_rate_timer;
+    // force triggering of groups
```
"This is used by "

andyp1per replied:
Fixed

andyp1per commented on ArduCopter/Copter.cpp:
```diff
@@ -259,6 +260,9 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
 #if HAL_BUTTON_ENABLED
     SCHED_TASK_CLASS(AP_Button,            &copter.button,              update,           5, 100, 168),
 #endif
+#if AP_INERTIALSENSOR_FAST_SAMPLE_WINDOW_ENABLED
```
add a comment - "don't delete"

andyp1per replied:
Fixed

andyp1per commented on ArduCopter/Parameters.cpp:
```diff
@@ -1232,6 +1233,22 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
+    AP_GROUPINFO("FSTRATE_ENABLE", 9, ParametersG2, att_enable, 0),
+
+    // @Param: FSTRATE_DIV
+    // @DisplayName: Fast rate thread divisor
```
Is there an easier way?

andyp1per replied:
add the value to the used value

andyp1per replied:
Fixed docs. Leave config as is

andyp1per commented on ArduCopter/motors.cpp:
```diff
@@ -181,7 +181,19 @@ void Copter::motors_output()
+// motors_output from main thread
+void Copter::motors_output_main()
+{
+    if (!using_rate_thread) {
```
how does this affect servos?

andyp1per replied:
Servos not affected

andyp1per commented on ArduCopter/motors.cpp:
```diff
@@ -181,7 +181,19 @@ void Copter::motors_output()
 
     // push all channels
-    SRV_Channels::push();
+    if (full_push) {
```
Needs a comment

andyp1per replied:
Fixed

rmackay9 commented on ArduCopter/motors.cpp:
```diff
@@ -181,7 +181,19 @@ void Copter::motors_output()
 
     // push all channels
-    SRV_Channels::push();
+    if (full_push) {
```
It's a shame that the copter code needs to worry about the detail of which push() to call.  The copter code essentially knows that the SRV_Channels cannot do what we want it to.  It would be nicer if the SRV_Channel::push() call could figure that out itself somehow

andyp1per replied:
Left as-is

rmackay9 commented on ArduCopter/rate_thread.cpp:
```diff
@@ -0,0 +1,501 @@
+
+        // set up rate thread requirements
+        if (!using_rate_thread) {
+            enable_fast_rate_loop(rate_decimation, rates);
```
not a blocker, feel free to resolve immediately but we need to be careful that we don't somehow end up in a loop where we continuously delete and re-create the buffer.

andyp1per replied:
Fixed

andyp1per commented on ArduCopter/rate_thread.cpp:
```diff
@@ -0,0 +1,501 @@
+// @Field: dtMin: Min time delta since last log output
+
+        if (now_ms - last_rtdt_log_ms >= 10) {    // 100 Hz
+            AP::logger().WriteStreaming("RTDT", "TimeUS,dt,dtAvg,dtMax,dtMin", "Qffff",
```
Do this the fast way

andyp1per replied:
Do we actually need this?

andyp1per replied:
Reduced rate to 10Hz and done the fast way

andyp1per commented on ArduCopter/rate_thread.cpp:
```diff
@@ -0,0 +1,501 @@
+    return 0;
+}
+
+void Copter::enable_fast_rate_loop(uint8_t rate_decimation, RateControllerRates& rates)
```
add some comments

andyp1per replied:
Fixed

andyp1per commented on ArduCopter/rate_thread.cpp:
```diff
@@ -0,0 +1,501 @@
+        motors_output(main_loop_count == 0);
+
+        // process filter updates
+        if (run_decimated_callback(rates.filter_rate, filter_loop_count)) {
```
Should this be called before the motors output or attitude controller?

andyp1per replied:
called from the right place

andyp1per commented on libraries/AP_HAL_ChibiOS/RCOutput.cpp:
```diff
@@ -1398,7 +1398,8 @@ void RCOutput::trigger_groups()
 #if !defined(HAL_NO_RCOUT_THREAD) || HAL_DSHOT_ENABLED
     // trigger a PWM send
-    if (!in_soft_serial() && hal.scheduler->in_main_thread() && rcout_thread_ctx) {
+    if (!in_soft_serial() &&
```
Fixed

peterbarker commented:
Ping @IamPete1 - good one to review while jet-lagged!

IamPete1 commented on libraries/AP_InertialSensor/FastRateBuffer.h:
```diff
@@ -0,0 +1,51 @@
+      binary semaphore for rate loop to use to start a rate loop when
+      we hav finished filtering the primary IMU
+     */
+    ObjectBuffer<Vector3f> _rate_loop_gyro_window{AP_INERTIAL_SENSOR_RATE_LOOP_BUFFER_SIZE};
```
Why do we need a buffer? I would have thought you could just keep the latest value and count how many times you update it, then zero the count when you read it.

andyp1per replied:
Still need to process a regular number of samples in the rate controller. I mean you could try and count and interpolate, but its simpler this way and simpler to understand without further changes to the rate controller code. Also in general the buffer only ever contains one item. We could consider reducing the size of the buffer.

IamPete1 commented on libraries/AP_InertialSensor/FastRateBuffer.h:
```diff
@@ -0,0 +1,51 @@
+    HAL_BinarySemaphore _notifier;
+    HAL_Semaphore _mutex;
+};
+#endif
```
```suggestion
#endif

```

andyp1per replied:
Fixed

IamPete1 reviewed (commented):
Have you tested turtle mode? It twiddles with stuff in motors.

Not a 100% review, still jet-lagged.

IamPete1 commented:
Has any testing been done without DShot? Does PWM still work I guess at it would still be at 400Hz. Do the DroneCAN ESC commands get sent faster with the higher update rate. Do the none motor outputs still work as expected (eg servo gimbal)?

andyp1per commented on ArduCopter/rate_thread.cpp:
```diff
@@ -0,0 +1,493 @@
+    rates.main_loop_rate = calc_gyro_decimation(rate_decimation, AP::scheduler().get_filtered_loop_rate_hz());
+    rates.filter_rate = calc_gyro_decimation(rate_decimation, ins.get_raw_gyro_rate_hz() / 2);
+
+    return 0;
```
Removed

andyp1per commented on ArduCopter/rate_thread.cpp:
```diff
@@ -0,0 +1,493 @@
+        const float dt = dt_us * 1.0e-6;
+        last_run_us = now_us;
+
+        motors->set_dt(sensor_dt);
```
I've moved it all to rate_controller_set_rates

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -53,6 +53,15 @@ class AP_InertialSensor_Backend
      */
     virtual bool update() = 0; /* front end */
 
+    // if AP_INERTIALSENSOR_FAST_SAMPLE_WINDOW_ENABLED
```
Yes, because of the vehicle include protections it can't actually be defined here, the comment is just to indicate that the code is technically controlled

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.h:
```diff
@@ -798,6 +803,29 @@ class AP_InertialSensor : AP_AccelCal_Client
     bool raw_logging_option_set(RAW_LOGGING_OPTION option) const {
         return (raw_logging_options.get() & int32_t(option)) != 0;
     }
+    // if AP_INERTIALSENSOR_FAST_SAMPLE_WINDOW_ENABLED
```
Yes, because of the vehicle include protections it can't actually be defined here, the comment is just to indicate that the code is technically controlled

andyp1per commented on ArduCopter/Parameters.cpp:
```diff
@@ -1232,6 +1233,22 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
+    // @Description: Fast rate thread divisor used to control the maximum fast rate update rate. The actual rate is the gyro rate in Hz divided by this value. This value is scaled depending on the configuration of FSTRATE_ENABLE.
+    // @User: Advanced
+    // @Range: 1 10
+    AP_GROUPINFO("FSTRATE_DIV", 10, ParametersG2, att_decimation, 1),
```
Yes, it will be downscaled on overload

andyp1per commented on ArduCopter/rate_thread.cpp:
```diff
@@ -0,0 +1,493 @@
+    const uint32_t attitude_rate = ins.get_raw_gyro_rate_hz() / rate_decimation;
+    attitude_control->set_notch_sample_rate(attitude_rate);
+    hal.rcout->set_dshot_rate(SRV_Channels::get_dshot_rate(), attitude_rate);
+    motors->set_dt(1.0f / attitude_rate);
```
Moved it here

andyp1per commented on libraries/AP_AHRS/AP_AHRS.h:
```diff
@@ -620,7 +620,7 @@ class AP_AHRS {
     // return primary accels, for lua
     const Vector3f &get_accel(void) const {
-        return AP::ins().get_accel();
+        return AP::ins().get_accel(_get_primary_accel_index());
```
I can remove it. It is correct however. What would you prefer?

IamPete1 replied:
A separate PR would be great, thanks.

andyp1per replied:
Ok done

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/MambaH743v4/hwdef.dat:
```diff
@@ -20,6 +20,7 @@ define CH_CFG_ST_RESOLUTION 16
 
 # leave 1 sectors free
 FLASH_RESERVE_START_KB 384
+env OPTIMIZE -O2
```
I think so, the PR is quite CPU intensive and I have been testing on this board a lot - this gives more headroom

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.h:
```diff
@@ -9,6 +9,8 @@
 #define AP_INERTIAL_SENSOR_ACCEL_VIBE_FILT_HZ           2.0f    // accel vibration filter hz
 #define AP_INERTIAL_SENSOR_ACCEL_PEAK_DETECT_TIMEOUT_MS 500     // peak-hold detector timeout
 
+#define AP_INERTIAL_SENSOR_RATE_LOOP_BUFFER_SIZE 8     // gyro buffer size for rate loop
```
Removed

andyp1per commented on ArduCopter/Copter.h:
```diff
@@ -725,7 +736,25 @@ class Copter : public AP_Vehicle {
-    void run_rate_controller();
+    void run_rate_controller_main();
+
+    // if AP_INERTIALSENSOR_FAST_SAMPLE_WINDOW_ENABLED
```
Yes, because of the vehicle include protections it can't actually be defined here, the comment is just to indicate that the code is technically controlled

andyp1per commented:
> Has any testing been done without DShot? Does PWM still work I guess at it would still be at 400Hz. Do the DroneCAN ESC commands get sent faster with the higher update rate. Do the none motor outputs still work as expected (eg servo gimbal)?

I will need to do more testing. I believe this is all fine (including turtle mode), but will test further.

tridge commented on ArduCopter/motors.cpp:
```diff
@@ -181,7 +182,21 @@ void Copter::motors_output()
-    SRV_Channels::push();
+    if (full_push) {
+        // motor output including servos and other updates that need to run at the main loop rate
+        SRV_Channels::push();
```
do we end up with an unpaired cork/push?

andyp1per replied:
No, all checks out. I have added in the check however.

tridge reviewed (commented):
some change?

tridge commented:
i'm happy with the system level and the AP_InertialSensor and AP_Vehicle changes, so from the point of view of plane which doesn't use this I'm happy with this PR
I have not properly reviewed the copter specific changes

IamPete1 commented on libraries/AP_InertialSensor/FastRateBuffer.cpp:
```diff
@@ -0,0 +1,147 @@
+// whether or not to push the current gyro sample
+bool AP_InertialSensor::is_rate_loop_gyro_enabled(uint8_t instance) const
+{
+    return fast_rate_buffer_enabled && fast_rate_buffer->use_rate_loop_gyro_samples() && instance == AP::ahrs().get_primary_gyro_index();
```
Should `fast_rate_buffer != nullptr` be in here?

andyp1per replied:
fixed

IamPete1 commented on libraries/AP_InertialSensor/FastRateBuffer.cpp:
```diff
@@ -0,0 +1,147 @@
+    if (fast_rate_buffer == nullptr) {
+        fast_rate_buffer = NEW_NOTHROW FastRateBuffer();
+    }
+    fast_rate_buffer_enabled = true;
```
Should we check the buffer was allocated before enabling ?
```suggestion
    fast_rate_buffer_enabled = fast_rate_buffer != nullptr;
```

andyp1per replied:
Fixed

IamPete1 commented on libraries/AP_InertialSensor/FastRateBuffer.cpp:
```diff
@@ -0,0 +1,147 @@
+        return false;
+    }
+
+    return fast_rate_buffer->get_next_gyro_sample(gyro);
```
nulprt check?

andyp1per replied:
Fixed

andyp1per commented on ArduCopter/Copter.cpp:
```diff
@@ -741,11 +754,24 @@ void Copter::one_hz_loop()
+    if (!started_rate_thread && get_fast_rate_type() != FastRateType::FAST_RATE_DISABLED) {
+        if (hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&Copter::rate_controller_thread, void),
+                                         "rate",
+                                         1536, AP_HAL::Scheduler::PRIORITY_RCOUT, 1)) {
```
Its still memory. This was the value I needed that gave a decent amout of headroom

andyp1per commented on ArduCopter/Copter.cpp:
```diff
@@ -741,11 +754,24 @@ void Copter::one_hz_loop()
+#if AP_INERTIALSENSOR_FAST_SAMPLE_WINDOW_ENABLED
+    // see if we should have a separate rate thread
+    if (!started_rate_thread && get_fast_rate_type() != FastRateType::FAST_RATE_DISABLED) {
+        if (hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&Copter::rate_controller_thread, void),
```
Fixed

andyp1per commented on ArduCopter/Attitude.cpp:
```diff
@@ -4,18 +4,21 @@
+    if (!using_rate_thread) {
+        motors->set_dt(last_loop_time_s);
+        // only run the rate controller if we are not using the rate thread
+        attitude_control->rate_controller_run();
```
I don't understand the comment

andyp1per replied:
This was addressed by the rate control threading PR

IamPete1 commented on ArduCopter/Parameters.cpp:
```diff
@@ -1232,6 +1233,22 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
+    // @Description: Enable the fast Rate thread. In the default case the fast rate divisor, which controls the update frequency of the thread, is dynamically scaled from FSTRATE_DIV to avoid overrun in the gyro sample buffer and main loop slow-downs. Other values can be selected to fix the divisor to FSTRATE_DIV on arming or always.
+    // @User: Advanced
+    // @Values: 0:Disabled,1:Enabled-Dynamic,2:Enabled-FixedWhenArmed,3:Enabled-Fixed
+    AP_GROUPINFO("FSTRATE_ENABLE", 9, ParametersG2, att_enable, 0),
```
Do you forsee more params here in the future? It would be nice to make this a enable in a little sub tree.

IamPete1 commented on ArduCopter/Copter.h:
```diff
@@ -631,6 +631,17 @@ class Copter : public AP_Vehicle {
+        FAST_RATE_FIXED               = 3,
+    };
+
+    FastRateType get_fast_rate_type() const { return FastRateType(g2.att_enable.get()); }
```
What happens if I set a out of range value? Lots of places are checking `!= FAST_RATE_DISABLED` what would a values of -1 result in? Unfortunately we have a presdent that -1 is the same as disabled for lots of params.

IamPete1 reviewed (approved):
I think this is as good enough to go in, I can't see any more issues from review, just need to start getting wider testing i think.

rafolg commented:
should HAL_INS_RATE_LOOP be INS_GYRO_RATE instead? I don't see where HAL_INS_RATE_LOOP is used?

rmackay9 reviewed (approved):
TBH, I have not reviewed this too thoroughly but I'm happy based on the review of others for this to go in
